src/main/java/org/reaktivity/nukleus/maven/plugin/internal/generate/EnumTypeGenerator.java

```diff
@@ -71,48 +93,64 @@ private NameConstantGenerator(
         }
 
         public NameConstantGenerator addValue(
-            String name)
+            String name, Integer value)
         {
-            builder.addEnumConstant(name);
+            if (value == null)
+            {
+                builder.addEnumConstant(name);
+            }
+            else
+            {
+                builder.addEnumConstant(name, TypeSpec.anonymousClassBuilder("$L", value).build());
+            }
             return this;
         }
     }
 
     private final class ValueOfMethodGenerator extends MethodSpecGenerator
     {
         private final List<String> values = new LinkedList<>();
+        private final Map<String, Integer> valueAssociatedWithConstant = new HashMap<>();
 
         private ValueOfMethodGenerator(
             ClassName enumName)
         {
             super(methodBuilder("valueOf")
                     .addModifiers(PUBLIC, STATIC)
-                    .addParameter(int.class, "ordinal")
                     .returns(enumName));
         }
 
         public ValueOfMethodGenerator addValue(
-            String name)
+            String name, Integer value)
         {
             values.add(name);
+            if (value != null)
+            {
+                valueAssociatedWithConstant.put(name, value);
+            }
             return this;
         }
 
         @Override
         public MethodSpec generate()
         {
-            builder.beginControlFlow("switch ($L)", "ordinal");
+            builder.addParameter(int.class, valueType == null ? "ordinal" : "value");
+            builder.beginControlFlow("switch ($L)", valueType == null ? "ordinal" : "value");
 
             for (int index=0; index < values.size(); index++)
             {
-                builder.beginControlFlow("case $L:", index)
-                       .addStatement("return $N", values.get(index))
+                String enumConstant = values.get(index);
+                int kind = valueAssociatedWithConstant.get(enumConstant) == null ? index :
+                    valueAssociatedWithConstant.get(enumConstant);
+                builder.beginControlFlow("case $L:", kind)
+                       .addStatement("return $N", enumConstant)
                        .endControlFlow();
             }
 
-            builder.endControlFlow()
-                   .addStatement("throw new IllegalArgumentException(String.format($S, ordinal))",
-                           "Unrecognized value: %d");
+            builder.endControlFlow().addStatement(valueType == null ?
+                "throw new IllegalArgumentException(String.format($S, ordinal))" :
+                "throw new IllegalArgumentException(String.format($S, value))",
+                valueType == null ? "Unrecognized ordinal: %d" : "Unrecognized value: %d");
```

Comment:
Use `String.format(...)` to create the main code line, parameterized with `%s`, filled by either `"ordinal"` or `"value"` depending on `valueType == null`.